File: 2020/3/main.py
from functools import reduce

def read_input():
    with open('input.txt') as input_file:
        input = input_file.read()

    return input.splitlines()

input = read_input()

length = len(input[0])
height = len(input)
tree = '#'
tree_count = 0

variants = [
    (1, 1),
    (3, 1),
    (5, 1),
    (7, 1),
    (1, 2),  # ?
]

total = []

for right, down in variants:
    tree_count = 0
    row, column = 0, 0
    while True:
        row += right
        column += down
        if row > (length - 1):
            row = row - length
        if column >= height:
            break
        if input[column][row] is tree:
            tree_count += 1

    print(tree_count)
    total.append(tree_count)

print(reduce(lambda x, y: x*y, total))

File: 2020/3/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func read_input() []string {
	input, _ := os.ReadFile("input.txt")
	s := string(input)
	split := strings.Split(s, "\n")
	return split[:len(split)-1]
}

func main() {

	var totals int
	totals = 1

	type Coordinates struct {
		X int
		Y int
	}

	input := read_input()
	var variants = []Coordinates{
		{X: 1, Y: 1},
		{X: 3, Y: 1},
		{X: 5, Y: 1},
		{X: 7, Y: 1},
		{X: 1, Y: 2},
	}

	for _, coord := range variants {
		tree_count := 0
		row := 0
		column := 0
		max_file := len(input)
		max_line := len(input[0])

		for column < max_file {
			// fmt.Println(string(input[column][row]))
			if string(input[column][row]) == "#" {
				tree_count += 1
			}

			row += coord.X
			column += coord.Y
			if row >= max_line {
				row = row - max_line
			}
		}
		fmt.Println(tree_count)
		totals = totals * tree_count
	}

	fmt.Println(totals)
}
